### PR TITLE
solution for the simple calculator challenge

### PR DESCRIPTION
**Pull request by Mo-Altayeb**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by ObayCipher** [comment]
Change the part of the function in test file from nodule for basic_calculator to module for simple_calculator function.

**Comment by ObayCipher** [comment]
**Make them pass in simple_calculator.py**:

1. It's better to explicitly state that the function returns None when an unsupported operation is passed.
2. No Assertions for Unsupported Operations:
3. It's better if an unsupported operation is passed, it returns None but doesn't raise an exception or provide a clear error message. A ValueError or similar exception.
4. Raising Exceptions for Invalid Operations:
Assertion that used to devided by zero should  provide an **error message**
5. Test for Floating Point Division:
Mentions it  in docstring, but it's not always clear from the code itself.

**Comment by ObayCipher** [comment]
**test_simple_calculator.py**

1. Typo in test_division_by_zero (wit h instead of with).

2. Some of the docstrings are quite short and could be expanded for clarity.

3. Floating-point precision isn't always checked in division tests for that make it in 3 decimal.

4. Edge case for the None return when unsupported operations are used is missing.

**Comment by Mo-Altayeb** [comment]
Made all the changes, what do you think?😁

File: solutions/simple_calculator.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# -*- coding: utf-8 -*-
+"""
+A module for performing multiplication, division, summation, or subtraction
+operations between two numbers.
+
+Module contents:
+    - simple_calculator: performs basic operations on two integers .
+
+Created on 5/1/2024
+@author: Mohamed Altayeb
```
Comment (ObayCipher) [inline solutions/simple_calculator.py:11]: The header for the function in the docstring should be in this structure:

```python
"""
A module for performing multiplication, division, summation, or subtraction
operations between two numbers.

Created on 5/1/2024
@author: Mohamed Altayeb
Group: ET foundations group 16 (Matrix)
    
Functions:
This module performs basic operations on two integers .
```

Comment (Mo-Altayeb) [reply solutions/simple_calculator.py:11]: working on it

Comment (Mo-Altayeb) [reply solutions/simple_calculator.py:11]: working on it

Comment (Mo-Altayeb) [reply solutions/simple_calculator.py:8]: Done! the function now matches the template 💯

Comment (ObayCipher) [reply solutions/simple_calculator.py:8]: Great work Nice, Clear and Clean Code :thumbsup:

**Review by ObayCipher — APPROVED** [review]
Great work Nice, Clear and Clean Code :thumbsup:

File: solutions/simple_calculator.py (hunk 1 above)
Comment (ObayCipher) [reply solutions/simple_calculator.py:8]: fixed
